### PR TITLE
Bugfixes + property goto

### PR DESCRIPTION
**Pull request by CWDN**
##### Features
* Added property goto. Works very similar to how functions works (almost the exact same code)
* Added a notification if the function goto isn't able to jump because proxy.methods returns an error.
 * Found this when trying to jump to an Eloquent function in Laravel.

##### Bugfixes (I hope)
* Fixed if the goto class is in the same namespace as the currently opened class
* Fixed the namespace and class selectors for alt-click not always working.
* Re-enabled the class goto. (From testing I couldn't replicate any problems)
* Fixed getStackClasses not working with things like 
```
$test = parent::foo()
                ---
```
```
$test->foo(function($test) {
       return $test;
})->filter()
    ------
```

* Did some updates to findFullWord function as I needed to define between a property and a function using `(`
* Fixed if the phpdoc has a return of static it not return the methods of itself.
 * Found this in Laravel's Collection as the phpdoc of the funciton has `@return static`

##### Ideas (I realise they aren't related to the PR but I didn't want to raise 2 issues for this)
I'm just suggesting these as I don't want to program anything you don't want in your package. 

* Refactor rename of property, function, class name.
 * So when you perform some action (keyboard shortcut, right click...etc) it will bring up a panel that allows you to rename a variable, property...etc. It will then update all references to the new name. [Youtube example](https://www.youtube.com/watch?v=iNJNdm5azdo)
* Have a repository where we can test this package on. It then means people can PR their code into this  test repository and we/you can properly debug it and find out why something isn't working.
 * A bit like what @hotoiledgoblinsack did when trying to explain a bug he found. [Link](https://github.com/Peekmo/atom-autocomplete-php/issues/52#issuecomment-128775877)

### CODE REVIEW

File: lib/services/php-file-parser.coffee [hunk 1]
```diff
@@ -273,6 +280,8 @@ module.exports =
       element = element.replace /^\s+|\s+$/g, ""
       if element[0] == '{' or element[0] == '(' or element[0] == '['
         element = element.substring(1)
+      else if element.indexOf('return ') == 0
```
Comment (Peekmo) [inline lib/services/php-file-parser.coffee]: I don't understand why you added this case ?

**Comment by Peekmo** [comment]
Hello,

Thanks for the PR.
I just have one question (see before) and for me, nothing is working better for something like that
```php
$test = parent::foo()
```
What should work ?

I didn't reproduce the problem with the GOTO that makes me add a return on top of the function.
The problem was that when I was in the pane opened by the GOTO, and then selecting text, the goto was fired, and so a lot of panes were opened.

An other point, is about your jumpToProperty and jumpToFunction methods. I didn't try to implement this, but in your opinion, will it be possible to use PHP reflection method "getStartLine()" (http://php.net/manual/en/reflectionfunctionabstract.getstartline.php) instead of using your regex to find the words in the file ? I think that it will consume less memory than regex if the file is a bit big.

For the refactor, why not, it's a really big feature that needs thinking to be efficient. But my major problem today, before that (because it's needed to have a powerfull refactoring), is to find a way to get that kind of autocomplete : 
```php
$x = new \DateTime();
$x->{here}

//or

$x = self::myfunction();
$x->{here}
```
In other words, I need getStackClasses() function to be able to parse results from other providers. Perhaps should I not use multiple providers ? I don't have any idea on how to be able to do that. And if you have a piece of idea, you're welcome !

For the test repository, it could be a nice idea to provide a repository that contains a lot of examples of what is working, and if a bug is found, someone can fork the repository and add its problem. In fact, great idea !

Thank you for your help :+1: 
Peekmo

**Comment by hotoiledgoblinsack** [comment]
I actually had the same idea about the repository yesterday. I had some testing code lying around and uploaded it to a repository here [1]. Feel free to fork it, use it, or, if you want, I can transfer ownership.

I can also confirm this fixes #56. However, I found one (minor) issue, if I do this:

```
/**
 * Tests absolute namespaces.
 *
 * @return \TestNamespace\Bar\BarClass
 *
 * @throws Test
 */
```

... holding alt over the return class freaks out and shows me a line that goes all the way to the line underneath it.

One other thing I noticed, not sure if this is intended or not and it may have been present before, is that there is no autocompletion if a word is already present, e.g.:

```
$this->{get autocompletion} // Works
$this->foo(); // Remove and readd the >, no autocompletion because foo is already present.
```

[1] https://github.com/hotoiledgoblinsack/php-autocomplete-test

File: lib/services/php-file-parser.coffee (hunk 1 above)
Comment (CWDN) [inline lib/services/php-file-parser.coffee]: I'm using getStackClasses on every type of function call in the function goto. This means that both static and local calls are being passed. `getStackClasses` wasn't coping with there not being a $ sign to stop the backwards search. So when you have a function call that isn't called from a $ variable (e.g. $this) then it keeps going backwards up the editor looking for the $. If it hits a `{` it stops and returns from that buffer position to the starting buffer position.

Example:
```
{

        return parent::foo();
```

I adjusted it so that it returned when the next character it's going to look at is `{`.

It then returned:
```

      return parent::foo();
```

Finally I trim the text and return that. Now when it comes to parsing the stack classes the return is still in there. So line 283 and 284 are removing that return if it is found.

**Comment by CWDN** [comment]
@Peekmo 
I've refactored the jumpTo into a single function in the abstract-goto that is now used by all three gotos.

Also I've used the getStartLine(), but it's only available for functions and not properties. So I'm still having to parse the entire editor for properties.

I don't understand what you are trying to say here, would you be able to give an example?:

> What should work ?
>
> I didn't reproduce the problem with the GOTO that makes me add a return on top of the function.
> The problem was that when I was in the pane opened by the GOTO, and then selecting text, the goto was fired, and so a lot of panes were opened.

@hotoiledgoblinsack 
Thanks for creating a new repository, shall we look at splitting that repository up into the different frameworks out there, like Laravel, Symfony and Zend? Then maybe a free area with the test code you currently have in there.

I don't seem to be able to replicate your bug with the goto. Seems to work perfectly fine for me. See below:

screenshot from 2015-08-15 00-14-24

**Comment by Peekmo** [comment]
Hello,

Very nice, thank you :+1:

**Comment by hotoiledgoblinsack** [comment]
@CWDN I just updated to the new version containing this pull request and also can't reproduce it anymore (previously I applied the patch from this pull request). Feel free to use the code in the repository however you wish (I quickly added the MIT license just to have some form of license as when it's omitted the licensing can differ from country to country), I just quickly built this to test issues I have as (sadly) the code I work with every day is proprietary.
